tests/fixtures/expanders.php

```diff
@@ -13,3 +13,63 @@ function my_hello_expander(TokenStream $ts) : TokenStream {
         )
     ;
 }
+
+function reverse_ast_expander(ParsedTokenStream $stream, Engine $engine) : ParsedTokenStream {
+    $ast = $stream->getAst();
+
+    $tokens = $ast->tokens();
+    $first = array_shift($tokens);
+
+    _modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
+        $value["outer"]["inner"][0] = new Token($first->type(), strrev($first->value()));
+        return $value;
+    });
+
+    return ParsedTokenStream::fromSource(
+        join(" ", $ast->tokens())
+    );
+}
+
+function upper_ast_expander(ParsedTokenStream $stream, Engine $engine) : ParsedTokenStream {
+    $ast = $stream->getAst();
+
+    $tokens = $ast->tokens();
+    $first = array_shift($tokens);
+
+    _modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
+        $value["outer"]["inner"][0] = new Token($first->type(), strtoupper($first->value()));
+        return $value;
+    });
+
+    return ParsedTokenStream::fromSource(
+        join(" ",  $ast->tokens())
+    );
+}
+
+function wrap_ast_expander(ParsedTokenStream $stream, Engine $engine) : ParsedTokenStream {
+    $ast = $stream->getAst();
+
+    $tokens = $ast->tokens();
+    $first = array_shift($tokens);
+
+    _modifyPrivateProperty($ast, "ast", function ($value) use ($first) {
+        $value["outer"]["inner"][0] = new Token($first->type(), "[" . $first->value() . "]");
+        return $value;
+    });
+
+    return ParsedTokenStream::fromSource(
+        join(" ",  $ast->tokens())
+    );
```

Comment:
> Curious side-effect of this is that it seems like; once you start dealing with the AST directly, you're likely to continue to use and modify the AST. Returning a token stream becomes almost optional. Except for the case where you don't wrap expanders in other expanders.

Still, I think it's kinda bad, when you only care about AST, to possibly have a token stream that is no longer in "sync". I feel like we should find a way to make the `ParsedTokenStream` to be rebuilt automatically outside the expander based on the AST. I'll think better about this.

Reply:
It's not great, but I think it's ok from the point of view that nested expanders need access to both the match AST and the replace token stream. This test is a bad example, because all the expanders want the same match data, whereas they could only be interested in the replaced tokens...